File: Introduction_to_OpenMP/Solutions/ftn/src/pi-omp-v1.f90
! Calculating Pi with using method: 
!     integral 4.0/(1+x^2) dx = p	     
! 	   OpenMP Version 1	     
program pi_omp_v1
implicit none
integer(kind=8) :: num_steps = 100000000000
integer(kind=8) i,tid,nthreads,omp_get_num_threads,omp_get_thread_num
real*8 :: PI_ref=3.1415926535897932,step,x,pi
real*8 :: summation=0.0
real*8   :: start,finish,omp_get_wtime
!call cpu_time(start)
step = 1/(1.0 * num_steps)
start=omp_get_wtime()
!$omp parallel private(x,tid) firstprivate(summation) shared(pi,nthreads,step)
	tid=omp_get_thread_num();
!$omp single
	nthreads=omp_get_num_threads();
!$omp end single
	do i =tid,num_steps-1,nthreads
		x=((i-1)+0.5) * step;
		summation=summation + (4.0/(1.0+(x*x)))
	end do
!$omp atomic update 
	pi=pi+(step*summation)

!$omp end parallel 
!call cpu_time(finish)
finish=omp_get_wtime()
	write(*,*) "Number of threads= ",nthreads
	write(*,*) "Estimated value of pi =",pi
	write(*,*) "Error =", abs(PI_ref - pi)
	write(*,*) "Compute time=", (finish-start)," seconds"
end

File: Introduction_to_OpenMP/Solutions/c/src/pi-omp-v1.c
/********************************************************
 * Calculating Pi with using midpoint rectangle method: *
 *  integral 4.0/(1+x^2) dx = pi	     		*
 *  or 							*
 *  summation (for i=0 to n) 4.0/(1+x_i^2)              *
 *  	where x_i = (i + 1/2) * delta x			*
 * 	   OpenMP Version 1          			*
 ********************************************************/
#include <omp.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

double PI_ref=3.1415926535897932;

static long long num_rects=100000000000;
double step;
int main (int argc, char* argv[]){
	double start,stop,diff_time;

	long long i;
	start= omp_get_wtime();
	double x,
	       sum=0.0,pi=0.0;
	step = 1.0 / (double) num_rects;
	int nthreads=0;
#pragma omp parallel private(i,x) firstprivate(sum) shared(pi,nthreads)
	{	
		//All threads retrieve their thread id
		int tid= omp_get_thread_num();
//Any one of the threads retrieve the total number of threads in the shared variable nthreads
#pragma omp single 
		{

			nthreads = omp_get_num_threads();

		}
		//Partition the work into block cyclic manner store local sum in a private variable
		for (i=tid; i<num_rects; i+=nthreads){
			x= step * ((double) i+0.5);
			sum = sum+4.0/(1.0+(x*x));
		}	

		
		//Each thread does the following one at a time.
#pragma omp critical (summation)
		{
			pi += (step * sum);
		}
		

	}

	stop= omp_get_wtime();
	diff_time = (double) (stop-start);
	printf ("Estimated value of pi = %2.16lf\n",pi);
	printf ("Error = %2.16lf\n",fabs(PI_ref - pi));
	printf ("Compute time= %2.5lf seconds\n",diff_time);

}
